lib/bolt/inventory/group2.rb

```diff
@@ -2,166 +2,183 @@
+          validate_proc = proc { |*args| }
+        end
 
-            Concurrent::Delay.new do
-              begin
-                hook.call(value)
-              rescue StandardError => e
-                loc = "resolve_reference in #{@name}"
-                raise Bolt::Plugin::PluginError::ExecutionError.new(e.message, plugin_name, loc)
-              end
-            end
-          else
-            value
-          end
+        validate_proc.call(reference)
+
+        begin
+          # Evaluate the plugin and then recursively evaluate any plugin returned by it.
+          hook.call(reference)
+        rescue StandardError => e
+          loc = "resolve_reference in #{@name}"
+          raise Bolt::Plugin::PluginError::ExecutionError.new(e.message, plugin_name, loc)
         end
       end
-      private :config_only_plugin
+      private :resolve_single_reference
+
+      # Checks whether a given value is a _plugin reference
+      def reference?(input)
+        input.is_a?(Hash) && input.key?('_plugin')
+      end
 
       def target_data(target_name)
-        if (data = @targets[target_name])
-          { 'config' => data['config'] || {},
-            'vars' => data['vars'] || {},
-            'facts' => data['facts'] || {},
-            'features' => data['features'] || [],
-            'plugin_hooks' => data['plugin_hooks'] || {},
-            # This allows us to determine if a target was found?
-            'name' => data['name'] || nil,
-            'uri' => data['uri'] || nil,
+        if @unresolved_targets.key?(target_name)
+          target = @unresolved_targets.delete(target_name)
+          resolved_data = resolve_data_keys(target, target_name).merge(
+            'name' => target['name'],
+            'uri' => target['uri'],
             # groups come from group_data
-            'groups' => [] }
+            'groups' => []
+          )
+          @resolved_targets[target_name] = resolved_data
+        else
+          @resolved_targets[target_name]
         end
       end
 
-      def add_target(target)
+      def add_target_definition(target)
         # This check ensures target lookup plugins do not returns bare strings.
         # Remove it if we decide to allows task plugins to return string node
         # names.
         unless target.is_a?(Hash)
           raise ValidationError.new("Node entry must be a Hash, not #{target.class}", @name)
         end
 
-        # This check prevents plugins from returning plugins
-        raise ValidationError.new("Cannot set target with plugin", @name) if target.key?('_plugin')
-        target.each do |k, v|
-          next if k == 'config'
-          validate_config_plugin(v, k, @name)
-        end
+        target['name'] = resolve_references(target['name']) if reference?(target['name'])
+        target['uri'] = resolve_references(target['uri']) if reference?(target['uri'])
+        target['alias'] = resolve_references(target['alias']) if reference?(target['alias'])
```

Comment:
couldn't this contain a reference?

`['mynode', {_plugin: myalias}]`

Reply:
You're right, it could. I think we can actually just remove the `if reference?` checks now since `resolve_references` has that check embedded.

Reply:
I updated these to check for the presence of the key rather than whether it's a reference.